File: runinator-command-center-flutter/lib/core/realtime/reconnect_backoff.dart
// shared exponential-backoff-with-jitter helper for websocket reconnects.
//
// a fixed reconnect delay means every client that drops at the same moment (a
// server restart, a rate-limit blip that closes sockets) reconnects in lockstep,
// which regenerates the exact spike that caused the drop. this staggers retries
// across clients and backs off on repeated failures instead of hammering at a
// constant interval forever.

import 'dart:math';

class ReconnectBackoff {
  ReconnectBackoff({
    this.initial = const Duration(milliseconds: 2000),
    this.max = const Duration(seconds: 30),
    this.factor = 1.7,
    Random? random,
  }) : _random = random ?? Random();

  final Duration initial;
  final Duration max;
  final double factor;
  final Random _random;

  int _attempt = 0;

  /// next delay to wait before reconnecting; grows with repeated failures and
  /// is jittered so concurrent clients don't retry in lockstep.
  Duration next() {
    final initialMs = initial.inMilliseconds.toDouble();
    final maxMs = max.inMilliseconds.toDouble();
    final rawMs = (initialMs * pow(factor, _attempt)).clamp(initialMs, maxMs);
    _attempt += 1;

    // full jitter within [initialMs/2, rawMs], so retries spread out immediately
    // rather than only once the backoff has grown.
    final floorMs = initialMs / 2;
    final jitteredMs = floorMs + _random.nextDouble() * (rawMs - floorMs);
    return Duration(milliseconds: jitteredMs.round());
  }

  /// call after a successful connection so the next drop starts from scratch.
  void reset() {
    _attempt = 0;
  }
}

File: runinator-command-center/src/core/realtime/reconnect-backoff.ts
// exponential backoff with jitter for websocket reconnects.
//
// a fixed reconnect delay means every client that drops at the same moment (a
// server restart, a rate-limit blip that closes sockets) reconnects in lockstep,
// which regenerates the exact spike that caused the drop. this staggers retries
// across clients and backs off on repeated failures instead of hammering at a
// constant interval forever.

export interface ReconnectBackoffOptions {
  initialMs?: number;
  maxMs?: number;
  factor?: number;
  random?: () => number;
}

export class ReconnectBackoff {
  private readonly initialMs: number;
  private readonly maxMs: number;
  private readonly factor: number;
  private readonly random: () => number;
  private attempt = 0;

  constructor(options: ReconnectBackoffOptions = {}) {
    this.initialMs = options.initialMs ?? 2000;
    this.maxMs = options.maxMs ?? 30000;
    this.factor = options.factor ?? 1.7;
    this.random = options.random ?? Math.random;
  }

  // next delay to wait before reconnecting; grows with repeated failures and
  // is jittered so concurrent clients don't retry in lockstep.
  next(): number {
    const rawMs = Math.min(Math.max(this.initialMs * this.factor ** this.attempt, this.initialMs), this.maxMs);
    this.attempt += 1;

    // full jitter within [initialMs/2, rawMs], so retries spread out immediately
    // rather than only once the backoff has grown.
    const floorMs = this.initialMs / 2;
    return Math.round(floorMs + this.random() * (rawMs - floorMs));
  }

  // call after a successful connection so the next drop starts from scratch.
  reset(): void {
    this.attempt = 0;
  }
}
